From 20dac9e6b4a684c311bbdfa8c65b01b6ee013672 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:37:28 -0400
Subject: [ethernet] i40e: report VF link state correctly

Message-id: <1409658042-22689-88-git-send-email-sassmann@redhat.com>
Patchwork-id: 90284
O-Subject: [RHEL7.1 PATCH 087/281] i40e: report VF link state correctly
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Mitch Williams <mitch.a.williams@intel.com>

Although the driver would correctly allow the VF link state to be
controlled by 'ip set link', it would not report it correctly back.

Fix this by filling out the appropriate field in the vf info struct.

Change-ID: I58d8e356438190e1ee9660b424301af6f416cdbe
Signed-off-by: Mitch Williams <mitch.a.williams@intel.com>
Signed-off-by: Catherine Sullivan <catherine.sullivan@intel.com>
Tested-by: Sibai Li <sibai.li@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 84ca55a04fe8a445328f914b98bc8c34f43810f3)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c b/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
index 5421714..b27b2f5 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
@@ -2291,6 +2291,13 @@ int i40e_ndo_get_vf_config(struct net_device *netdev,
  ivi->vlan = le16_to_cpu(vsi->info.pvid) & I40E_VLAN_MASK;
  ivi->qos = (le16_to_cpu(vsi->info.pvid) & I40E_PRIORITY_MASK) >>
      I40E_VLAN_PRIORITY_SHIFT;
+ if (vf->link_forced == false)
+  ivi->linkstate = IFLA_VF_LINK_STATE_AUTO;
+ else if (vf->link_up == true)
+  ivi->linkstate = IFLA_VF_LINK_STATE_ENABLE;
+ else
+  ivi->linkstate = IFLA_VF_LINK_STATE_DISABLE;
+
  ret = 0;
 
 error_param:
-- 
1.7.1